src/python/BuildingControlsSimulator/DataClients/WeatherChannel.py

```diff
@@ -703,3 +705,245 @@ def to_epw(
+        
+        # Unlike the gridded WTK data the NSRDB is provided as sparse time-series dataset.
+        # The quickest way to find the nearest site it using a KDtree
+        def nearest_site(tree, lat_coord, lon_coord):
+            lat_lon = np.array([lat_coord, lon_coord])
+            dist, pos = tree.query(lat_lon)
+            return pos
+
+        # Open the desired year of nsrdb data
+        # server endpoint, username, password is found via a config file
+        f = h5pyd.File("/nrel/nsrdb/v3/nsrdb_{}.h5".format(str(year)), 'r')
+        
+        #create binary tree of coords for search
+        dset_coords = f['coordinates'][...]
+        tree = cKDTree(dset_coords)
+        
+        #identify nearest weather station
+        location_idx = nearest_site(tree, lat, long)
+        
+        strPath = '' #TODO:  Create environment variable for nsrdb cache folder
+        strFile = 'nsrdb_{0}_{1:.2f}_{2:.2f}.csv.gz'.format(str(year),dset_coords[location_idx][0],dset_coords[location_idx][1])
+
+        if not os.path.exists(strPath + strFile):
+            print('Pulling nsrdb data')
+            # Extract datetime index for datasets
+            time_index = pd.to_datetime(f['time_index'][...].astype(str), utc=True)# Temporal resolution is 30min
+
+            #get data sets
+            dset_dni  = f['dni']
+            dset_ghi  = f['ghi']
+            dset_temp = f['air_temperature']
+            dset_rh   = f['relative_humidity']
+            # list(f)  # for full list the datasets in the file
+
+            #filter on time series for given location
+            tseries_dni  = dset_dni[:,  location_idx] / dset_dni.attrs['psm_scale_factor']  
+            tseries_ghi  = dset_ghi[:,  location_idx] / dset_ghi.attrs['psm_scale_factor']  
+            tseries_temp = dset_temp[:, location_idx] / dset_temp.attrs['psm_scale_factor'] 
+            tseries_rh   = dset_rh[:,   location_idx] / dset_rh.attrs['psm_scale_factor']   
+
+            #combine into single DF int
+            df_solar = pd.DataFrame()
+            df_solar["datetime"]          = time_index   # datetime64
+            df_solar["dni"]               = tseries_dni  # W/m2, float
+            df_solar["ghi"]               = tseries_ghi  # W/m2, float
+            df_solar["temp_air"]          = tseries_temp # degC, float
+            df_solar["relative_humidity"] = tseries_rh   # %rh,  float
+
+            df_solar.to_csv(strPath + strFile, index=False, compression='gzip')
+        else:
+            print('Re-opening nsrdb data')
+            df_solar = pd.read_csv(strPath + strFile, compression='gzip')
+            df_solar.datetime = pd.to_datetime(df_solar.datetime)
+
+        if not df_solar.empty:
+            return df_solar
+        else:
+            logger.error("failed to retrieve nsrdb fill data.")
+            return 
+
+    #fill_nsrdb_data = attr.ib(default=None)
+    def fill_nsrdb(self, input_epw_data, datetime_channel, fill_nsrdb_data, sim_config):
```

Comment:
We have references and functionality specific to .epw data here that should be removed. The filling of the NSRDB data is independent of the .epw file that gets generated later in the simulation process during the generation of the EnergyPlus simulation files.

Reply:
actually now that i'm more familiar with this all, I noticed there was a lot of EPW stuff.  updated.